### PR TITLE
*: make string in select result convert to bit correctly.

### PR DESCRIPTION
**Pull request by bobotu**
```SQL
CREATE TABLE t (a BIT(64));
CREATE TABLE t1 (a VARCHAR(20));
INSERT INTO t1 VALUES ('10');
INSERT INTO t SELECT a FROM t1;
SELECT a+0 FROM t;
```
should return `12592`, but current TiDB return `10`.

In #3115 I ignored this case, and I fix it in this PR.

Due to the change of `Unflatten` I must change many old test case to adjust the type change, but I think this change may make future code handle string type easier.

### CODE REVIEW

**Comment by bobotu** [comment]
PTAL @coocood @shenli

File: ddl/ddl_db_test.go [hunk 1]
```diff
@@ -1104,4 +1104,11 @@ func (s *testDBSuite) TestIssue2858And2717(c *C) {
 	s.tk.MustExec("insert into t_issue_2858_hex values (123), (0x321)")
 	s.tk.MustQuery("select a from t_issue_2858_hex").Check(testkit.Rows("291", "123", "801"))
 	s.tk.MustExec(`alter table t_issue_2858_hex alter column a set default 0x321`)
+
+	s.tk.MustExec("drop table if exists t, t1")
+	s.tk.MustExec("create table t (a bit(64))")
+	s.tk.MustExec("create table t1 (a varchar(20))")
+	s.tk.MustExec(`insert t1 value ('10')`)
+	s.tk.MustExec(`insert t select a from t1`)
+	s.tk.MustQuery("select a+0 from t").Check(testkit.Rows("12592"))
```
Comment (coocood) [inline ddl/ddl_db_test.go]: I think this test should be put in `executor` package.

File: tablecodec/tablecodec.go [hunk 2]
```diff
@@ -388,10 +388,12 @@ func Unflatten(datum types.Datum, ft *types.FieldType, inIndex bool) (types.Datu
 	case mysql.TypeFloat:
 		datum.SetFloat32(float32(datum.GetFloat64()))
 		return datum, nil
+	case mysql.TypeVarchar, mysql.TypeString:
+		datum.SetString(datum.GetString())
```
Comment (coocood) [inline tablecodec/tablecodec.go]: `TypeString` and `TypeVarchar` not always be string type.
`[]byte` and `string` should be used interchangeable.
The difference is charset.
The binary types have `binary` charset, string types have `utf8` charset.

Comment (coocood) [inline tablecodec/tablecodec.go]: The datum kind String or Bytes should not affect the result of the execution.

Comment (coocood) [inline tablecodec/tablecodec.go]: Many places of the existing code don't handle it correctly. Maybe we should just remove `KindString`, only use `KindBytes` in the future.

Comment (bobotu) [inline tablecodec/tablecodec.go]: Yes, you are right. After check the MySQL `Data Types Reference`, I realized that the `Collate` and `Charset` is the main difference between `CHAR` and `BINARY`. 

But it seems like that `Unflatten` doesn't care the `Collate` defined in `FieldType`. That result in functions which need take different strategy to handle binary string and nonbinary string are difficult to distinguish them. I think we can set `Datum.k` and `Datum.collation` in `Unflatten` based on `Collate` and `Charset` in `FieldType`.
This change let functions easy to distinguish binary string and nonbinary string, while still let functions to judge what kind of datum based on charset and collate.

For bug mentioned in this PR, I finally find that MySQL treat all string as binary string when convert string to bit. So I revert previous commit and fix this bug in a new commit.

Comment (coocood) [inline tablecodec/tablecodec.go]: @bobotu 
Good idea!

**Comment by coocood** [comment]
LGTM

**Pull request merged**